egui_demo_lib/src/apps/demo/view_edit.rs

```diff
@@ -0,0 +1,96 @@
+//! Source code example about creating other type of your widget which uses `egui::Memory` and
+//! created using a combination of existing widgets.
+//! This is meant to be read as a tutorial, hence the plethora of comments.
+
+use egui::Id;
+use std::hash::Hash;
+
+/// Easymarkup text editor with the ability to preview a result.
+///
+/// ## Example:
+/// ``` ignore
+/// toggle_ui(ui, &mut my_text, "description_1");
+/// ```
+pub fn view_edit_ui(ui: &mut egui::Ui, text: &mut String, id_source: impl Hash) -> egui::Response {
+    // This widget has its own state - `View` or `Edit`,
+    // so there is the algorithm for type of widgets:
+    //  1. Declare state struct
+    //  2. Create id
+    //  3. Get state for this widget
+    //  4. Process ui, change a local copy of the state
+    //  5. Insert changed state back
+
+    // 1. Declare state struct
+    // This struct represents the state of this widget. It must implement at least `Clone` and be
+    // `'static`. If you use the `persistence` feature, it also must implement
+    // `serde::{Deserialize, Serialize}`. You should prefer creating custom newtype structs
+    // or enums like this, to avoid TypeId intersection errors, especially when you use
+    // `Memory::data` without `Id`.
+    #[derive(Clone, Copy, Eq, PartialEq, Debug)]
+    #[cfg_attr(feature = "persistence", derive(serde::Deserialize, serde::Serialize))]
+    enum State {
+        View,
+        Edit,
+    }
+
+    // The default state will be set for first call of widget with this id.
+    impl Default for State {
+        fn default() -> Self {
+            State::View
+        }
+    }
+
+    // 2. Create id
+    let id = Id::new(id_source);
```

Comment:
This hinges on that `id_source` is unique in the entire program. It is better to use `ui.make_persistent_id(id_source);`

Reply:
Or use `ui.make_position_id()` and there is no need for `id_source`, and it will work as long as the number of proceeding widgets in the same UI doesn't change.

Reply:
The best thing about contributing is that I better understand how to use the library. Thanks for `make_persistent_id`, now I use this. Previously I didn't know what this function is doing, I only understand that after reading [these comments](https://docs.rs/egui/0.11.0/src/egui/ui.rs.html#28-33), I think they should be added to this function documentation.

`make_position_id` is deprecated and says that `allocate_space` should be used. But I don't think allocation is needed for a composite widget without custom painting and etc.